pkg/apis/gcpfirewall/v1alpha1/types.go

```diff
@@ -0,0 +1,181 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Protocol defines network protocols supported for GCP firewall.
+type Protocol string
+
+// CIDR defines a IP block.
+// TODO(b/186120065) Modify the validation to include IPv6 CIDRs with FW 3.0 support.
+// +kubebuilder:validation:Pattern=`^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(/(3[0-2]|2[0-9]|1[0-9]|[0-9]))?$`
+type CIDR string
+
+const (
+	// ProtocolTCP is the TCP protocol.
+	ProtocolTCP Protocol = "TCP"
+	// ProtocolUDP is the UDP protocol.
+	ProtocolUDP Protocol = "UDP"
+	// ProtocolICMP is the ICMP protocol.
+	ProtocolICMP Protocol = "ICMP"
+)
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:shortName=gf
+
+// GCPFirewall describes a GCP firewall spec that can be used to configure GCE
+// firewalls. A GCPFirewallSpec will correspond 1:1 with a GCE firewall rule.
+type GCPFirewall struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the desired configuration for GCP firewall
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	Spec GCPFirewallSpec `json:"spec,omitempty"`
+
+	// Status is the runtime status of this GCP firewall
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	Status GCPFirewallStatus `json:"status,omitempty"`
+}
+
+// GCPFirewallSpec provides the specification of a GCPFirewall.
+// The firewall rule apply to the cluster associated targets (network tags or
+// secure tags) which are deduced by the controller. As a result, the specified
+// rule applies to ALL nodes and pods in the cluster.
+type GCPFirewallSpec struct {
+	// Name of the firewall rule that's configured in GCP.
+	// For classic GCP firewalls, this field needs to be unique. For FW 3.0,
+	// this field will be in the rule description.
+	Name string `json:"name"`
+
+	// Rule action of the firewall rule. Only allow action is supported. If not
+	// specified, defaults to ALLOW.
+	// +optional
+	// +kubebuilder:validation:Enum=ALLOW
+	// +kubebuilder:default=ALLOW
+	Action string `json:"action"`
```

Comment:
do you want to include a few string constant to capture the valid input here?

Reply:
Right now, ALLOW is only action possible. We might extend this in future to include DENY. This can be considered readonly for now as this field can be left empty.

Reply:
Updated.